### PR TITLE
Bugfix keep sorting

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves https://github.com/FriendsOfCake/search/issues/120

I wish there was a more automatic/smarter solution like the cakedc one had, but that would mean more of the logic needs to be moved to the component.
This is a minimal invasive approach to at least support it basically, and support it properly working.
Most importantly it does not include the page by default, which would only lead to big trouble.

### CODE REVIEW

**Comment by dereuromark** [comment]
I made a live example of that branch:
http://sandbox.dereuromark.de/sandbox/search-examples?search=a&status=1&page=8&sort=CountryRecords.iso3&direction=asc
And then make the "a" an "ab" search input content, which will bring you to
http://sandbox.dereuromark.de/sandbox/search-examples?search=ab&status=1&sort=CountryRecords.iso3&direction=asc
page is removed as expected.

**Comment by ADmad** [comment]
LGTM but would be nice to have some tests for this so that there are not regressions in future.

**Comment by dereuromark** [comment]
Tests added, also fixed up composer json dependencies.

File: src/Controller/Component/PrgComponent.php [hunk 1]
```diff
@@ -13,14 +14,15 @@ class PrgComponent extends Component
...
+     *   processed if `false` none. Default is ['index', 'lookup'].
      * - `queryStringToData` : Set query string as request data. Default `true`.
      *
      * @var array
      */
     protected $_defaultConfig = [
         'actions' => ['index', 'lookup'],
-        'queryStringToData' => true
+        'queryStringToData' => true,
+        'queryStringWhitelist' => ['sort', 'direction']
```
Comment (inoas) [inline src/Controller/Component/PrgComponent.php]: `,` as far as the current outlook on that issue is.

File: src/Controller/Component/PrgComponent.php [hunk 2]
```diff
@@ -31,9 +33,11 @@ class PrgComponent extends Component
      */
     public function startup()
     {
-        if ($this->_actionCheck()) {
-            return $this->conversion();
+        if (!$this->_actionCheck()) {
+            return null;
         }
+
+        return $this->conversion();
```
Comment (inoas) [inline src/Controller/Component/PrgComponent.php]: Does this work without the negative condition?
```php
public function startup()
{
    if ($this->_actionCheck()) {
        return $this->conversion();
    }

    return null;
}
```

**Comment by inoas** [comment]
So as a whole this would persist sort and direction, but not page? What about any other params?
*Why not just remove the page and persist any other query string?*

File: src/Controller/Component/PrgComponent.php [hunk 3]
```diff
@@ -13,14 +14,17 @@ class PrgComponent extends Component
      * ### Options
      * - `actions` : Action name(s) to use PRG for. You can pass a single action
      *   as string or multiple as array. If boolean `true` all actions will be
-     *   processsed if `false` none. Default is ['index', 'lookup'].
+     *   processed if `false` none. Default is ['index', 'lookup'].
      * - `queryStringToData` : Set query string as request data. Default `true`.
+     * - `queryStringWhitelist` : An array of whitelisted query strings to be kept.
```
Comment (inoas) [inline src/Controller/Component/PrgComponent.php]: query string keys

**Comment by dereuromark** [comment]
Because that would destroy it again, as those can be from the previous search and would then force-keep every search value even if removed from the post.

File: src/Controller/Component/PrgComponent.php (hunk 2 above)
Comment (dereuromark) [inline src/Controller/Component/PrgComponent.php]: Its called "return early".

File: src/Controller/Component/PrgComponent.php (hunk 1 above)
Comment (dereuromark) [inline src/Controller/Component/PrgComponent.php]: ?

File: src/Controller/Component/PrgComponent.php (hunk 3 above)
Comment (dereuromark) [inline src/Controller/Component/PrgComponent.php]: Too late.

File: src/Controller/Component/PrgComponent.php (hunk 2 above)
Comment (inoas) [inline src/Controller/Component/PrgComponent.php]: Both are? Can we remove the negation in the if condition?

**Comment by inoas** [comment]
@dereuromark If you paginate multiple tables on one single view (through cells maybe, and scopes) wouldn't those then get removed?
The snipplet I posted to the docs (that always persists page for any scope) just needs to opt-out the page param for any given scope (even no-scope):
See: https://github.com/FriendsOfCake/search/issues/120#issuecomment-247191204

**Comment by dereuromark** [comment]
Your snippet does not work at all. For the reasons mentioned on the ticket.
And you can do that still if you want to, but a good, non-complicated and sane default solution here is the better way to go IMO.

**Comment by inoas** [comment]
The solution should be opt-out not opt-in for persisting query strings.
You want to opt-out of persisting any page query string key-value pair as those can lead to 404 responses.

**Comment by dereuromark** [comment]
So far the solution before worked as well, and now works even better with mine by persisting some whitelisted ones.
Opting out does not work here. Of course that would be better but cannot be done properly due to the reasons mentioned already.

**Comment by dereuromark** [comment]
Waiting for merge and release so I can update my sandbox to use a tagged version here :)
http://sandbox.dereuromark.de/sandbox/search-examples
Feel free to add more advanced search examples to show the full power of the plugin :P

**Comment by dereuromark** [comment]
Should limit param also be kept to be in sync with Paginator whilelist?

**Comment by dereuromark** [comment]
Done, this should now be good to go.

**Comment by ADmad** [comment]
@dereuromark This can be merged once docblock description is added for `_filterParams()`

**Comment by dereuromark** [comment]
Pretty harsh there ;) Done.

**Comment by ADmad** [comment]
I had already commented "Description please." above the method few days ago.

**Pull request merged**